controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -213,7 +214,13 @@ func (r *BareMetalHostReconciler) Reconcile(ctx context.Context, request ctrl.Re
 		request:        request,
 		bmcCredsSecret: bmcCredsSecret,
 	}
-	prov, err := r.ProvisionerFactory(*host, *bmcCreds, info.publishEvent)
+
+	factory := r.ProvisionerFactory
+	if host.Spec.ExternallyProvisioned && !host.HasBMCDetails() {
+		factory = empty.New
+	}
+
```

Comment:
The previous implementation was a wrapper function with the main goal of keeping the decisional logic well defined in one place only (for maintenance reasons), and also to reflect the factory having a more dynamic behavior as the new umanaged hosts feature required an evaluation per host (so at every reconcile loop), in contrast with the original implementation where the decision was more static (at the beginning). Definition remained in the main.go I guess for the sake of simplicity - and it could have been moved into a dedicated struct type probably - but the selection still happened in the controller.

The problem of the excessive logging is somehow a side effect of such requirement, since in a standard environment potentially some hosts could use the default provisioner (ironic) while others could use the empty one - considered a completely different provisioner; in such case it could be useful to know which provider has been been used for a given host, and maybe the log traces could be reviewed.

Not sure if I've missed something, but the current proposed implementation instead looks like more as an override (for all the cases - test, demo, default), which is a different approach.

A more sophisticated implementation could be to split between at the startup across the demo/test/production factories, and keep the production one as a factory of factories (deciding between ironic or empty)

Reply:
That's discussed to some extend in the commit message/PR description. The empty provisioner isn't really a provisioner, it's just a piece of runtime logic put in there to stop the real provisioner choking because of being given no BMC details.
Arguably a better approach might be to make the existing provisioners cope gracefully with not being provided BMC details. I guess one way to do that might be to decide between ironic vs empty in the factory for the ironic provisoner.

Reply:
As far as a backportable short-term fix goes though, I still think this is the best approach.

Reply:
> I guess one way to do that might be to decide between ironic vs empty in the factory for the ironic provisoner

My proposal was in that direction (but considering the empty provisioner a valid one, which is the discussion point I guess). For me it would be fine to move it there then, as it could be more clear that it's going to be used to cover the unmanaged case, and the demo/fixture provisioner  previous behaviors will not overridden.

Reply:
As [discussed on slack](https://kubernetes.slack.com/archives/CHD49TLE7/p1616436038039800) I think this behavior should be independent of whether the host is `externallyProvisioned` or not